### PR TITLE
Fix leaked alias symbol independent type params in printback

### LINKED ISSUE #42079: Syntax error in emitted declaration's generic arguments
# Bug Report

### 🔎 Search Terms

syntax error declaration emit generics "Cannot find name" ts2304

### 🕗 Version & Regression Information

- This is syntactically incorrect output from tsc
- It is not a regression
- This is the behavior in every version I tried (3.3 - 4.2.0-dev.20201222)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/dev/bug-workbench/?ts=4.2.0-dev.20201222#code/PTAEAEBMFMGMBsCGAnRAXAlgewHYCgQJF54sB3AKQGcAuUAM2KujwLHHo3mh0QFtodPogw4AdGip40ATwAO0UAHVUcgDwBBAHygAvKADeeUCdA5oVNHQ14AvgG48otNGSNYigArJoANx5oAJI48KKiAObBACJwSKiYuACifBhohnZ40AAecljIabIKoJ4ArgBGobAqiOoAGgA0oACaeoa2OvrV6k06AGTFPv44QSFhOJE4MQgo6Ng4yamO2bn5DCU4sAk4DDhqACo6ABQo4XR7AJSgdKUVGFWq+zpGpqA+aCXI2wZmFlagJ6BbBlYLhLD9LNBIK16DhDt96FgsHQAIy2c72UyEADKMmGiCyoESyGQeVAolAMGm8TmoGgKTSAAtXCxlnk0jBGCV4GlzBDII4gA)

**Important:** The playground itself has a rendering bug in the display of the declaration.  It still looks invalid.  But it's invalid in another way in the truthful output.  I paste the truthful output below.

### 💻 Code

```ts
// @declaration
type Wrap<A> = {
    nest: A
};
interface PreventInliningInDeclarationEmit {
}
export type PublicWrap<X, Y = {}> = Wrap<Y> & PreventInliningInDeclarationEmit;
export function fn<T> (arg: T) : PublicWrap<T> {
    return { nest: arg }
}
const nested = fn({ foo: 1});   // Syntax Error in declaration emit here
export default nested;
```

### 🙁 Actual behavior

Emitted declaration has syntax error.  `Cannot find name 'T'. (ts2304)`

```ts
declare type Wrap<A> = {
    nest: A;
};
interface PreventInliningInDeclarationEmit {
}
export declare type PublicWrap<X, Y = {}> = Wrap<Y> & PreventInliningInDeclarationEmit;
export declare function fn<T>(arg: T): PublicWrap<T>;
declare const nested: PublicWrap<T, {}>;
export default nested;
```

### 🙂 Expected behavior

A valid declaration.

```ts
declare type Wrap<A> = {
    nest: A;
};
interface PreventInliningInDeclarationEmit {
}
export declare type PublicWrap<X, Y = {}> = Wrap<Y> & PreventInliningInDeclarationEmit;
export declare function fn<T>(arg: T): PublicWrap<T>;
declare const nested: PublicWrap<{
    foo: number;
}>;
export default nested;
```

### LINKED ISSUE #44727: Incorrect transpilation of generic type parameter
# Bug Report

I have a type with an unused generic type parameter.\*
This type annotates an **optional parameter property** in the constructor of a class.
Starting with Version 4.2.X of TypeScript, this class is incorrectly transpiled, so that TS code using the transpiled sources (.d.ts) does not transpile anymore.
The error was:
```
error TS2304: Cannot find name 'T'
```

No wonder it cannot be found, it was not in the original code (please see the minimal example below).

During my investigation I found that this issue occurs, when these prerequisites are met:
1. The given type has an unused generic parameter.
1. The given type represents a union type.
1. You have a class with a constructor argument using the given type.
1. The constructor argument is a parameter property.
1. The constructor argument is optional.

*\* (Tried something. Didn't work. Postponed it, but left the parameter for later use.)*
### 🔎 Search Terms

generics, transpilation, missing parameter

### 🕗 Version & Regression Information

- This changed between versions 4.1.X and 4.2.X

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=5#code/C4TwDgpgBAQgTgewNYQHYBVwQDzoHxQC8UA5AIYlQA+pARiQNwCwAUKwMYA2ZAzj1AFkQAYW58oAb1ZQZUdglQ9gcAK7tgCOAAo4EMgBMFnEFDJwA5gH4AXLEQoMWbGVQg8ASkkBfVj5ZA)

### 💻 Code

Before transpilation:
```ts
type BrokenType<T> = 'a' | 'b';

class MyClass {
    constructor(readonly arg?: BrokenType<any>) {}
}
```

After transpilation
```ts
declare type BrokenType<T> = 'a' | 'b';
declare class MyClass {
    readonly arg?: BrokenType<T> | undefined;
    constructor(arg?: BrokenType<T> | undefined); // <----- T should be any
}
```

### 🙁 Actual behavior

The code is transpiled to contain the generic parameter name, instead of the actual type that was passed as argument.

### 🙂 Expected behavior

I would expect the result to contain the original type as specified in the original code. The generic parameter is not known when consuming the type. This worked before version 4.2.X.

### PR DESCRIPTION
**Pull request by weswigham**
When we measure the variance of an alias, an `Independent` type parameter is an unused type parameter - for whatever reason, it has no effect on the comparability of the resulting type. What this often means is that that type parameter also does not play into the construction of the the target type _at all_. That, in turn, means that due to interning, you can have multiple seemingly unique references to a type, which all point at a single underlying type - usually the one first manufactured within the type alias declaration itself. That underlying type, in turn, has an alias from when it was first created, and that alias's type arguments may contain type parameters only available within the scope of the original alias declaration. Normally, this would disqualify use of the alias as a reference external to that declaration, but we failed to detect such a case as our assumption is that if we are doing printback using a type parameter, that it will be made available in scope (otherwise, how did we get it?). Since this type parameter comes from far afield, we can't safely do printback using it, _however_, in this fix, rather than issuing an error in such a circumstance (which would be unfortunate and hard to fix for users), I detect when this situation arises by checking the variance and visibility of the alias symbol type arguments, and substitute an unreachable `Independent` type argument with `unknown`, as it will not affect the resulting type's shape, thus creating a functioning declaration file.

This approach required checking the variance of some type aliases where we previously had not bothered (usually simply because the test in question had not been written to require it), and in turn exposed a crash-y problem we had calculating variance while handling recursive references to unions and intersections within type aliases (and recursive normalization thereof - see the types in `recursivelyExpandingUnionNoStackoverflow` as an example of a type which we crashed when fetching the variance on it). To fix _that_, #37910 (which just got merged into `master`) mostly handles it, however the size limit it originally imposed has to be removed, as we need the deeply nested type guarding part of `recursiveTypeRelatedTo` even for the smallest unions, as recursive type references can create situations where these can infinitely expand upon normalization. In addition, I've added an identity for unions and intersections with aliases into `getRecursionIdentity` to improve how quickly we recognize these structures as deeply nested (the time it took to calculate variances in `recursiveTypeReferences2` proved that needed).

TL;DR: We now print
```ts
declare const nested: PublicWrap<unknown, {}>;
```
rather than the erroneous
```ts
declare const nested: PublicWrap<T, {}>;
```
where `PublicWrap`'s first type parameter is `Independent`.

Fixes #42079
Fixes #44727

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot perf test this - since this affects caches, this should cause some change to perf, but I can't predict if it'll be good or bad in real world scenarios; the caching is noncritial to fixing this, but making the extra caching conditional would complicate this change a bit, so I'd like to avoid it if possible.

Also:
@typescript-bot test this
@typescript-bot run dt
@typescript-bot user test this
since this affects `Independent` type parameter printback, this could change some baselines which may capture our current behavior. I've minimized how often it occurs, so hopefully there's not many changes beyond some duplicate elaborations going away from the expanded caching.

And @typescript-bot pack this for good measure.

**Comment by weswigham** [comment]
The user baselines all look great excepting a few new excessive stack warnings when calculating variance for some types in `ts-toolbelt`. I'm wondering if those errors could also have been triggered prior to this PR with some slight modifications to force variance analysis of the same types. In any case, I'll look at it.

**Comment by weswigham** [comment]
@typescript-bot perf test this faster

**Comment by weswigham** [comment]
Yeah, those perf numbers look fine.

**Comment by sandersn** [comment]
Unfortunately, we never finished reviewing this PR. It is pretty old now, so I'm going to close it to reduce the number of open PRs.
